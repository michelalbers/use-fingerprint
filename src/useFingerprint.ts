import React from "react";
import Fingerprint2 from "fingerprintjs2";

export default function () {
  const [fingerprint, setFingerprint] = React.useState<string>(null);

  React.useEffect(() => {
    if (typeof window !== "object") return; // Run on the server side - ignore

    const setFingerprintFromComponents = () => {
      Fingerprint2.get((components) => {
        const fingerprint = Fingerprint2.x64hash128(
          components.map((c) => c.value).join(""),
          31
        );
        setFingerprint(fingerprint);
      });
    };

    // @ts-ignore
    if (window.requestIdleCallback) {
      // @ts-ignore
      window.requestIdleCallback(setFingerprintFromComponents);
    } else {
      setTimeout(setFingerprintFromComponents, 500);
    }
  }, []);

  return fingerprint;
}
