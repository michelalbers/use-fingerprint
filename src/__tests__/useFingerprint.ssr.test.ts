/**
 * @jest-environment node
 */

import Fingerprint2 from "fingerprintjs2";
import useFingerprint from "../useFingerprint";
import { renderHook, act } from "@testing-library/react-hooks";

describe("useFingerprint hook", () => {
  beforeEach(() => {
    Fingerprint2.get = jest
      .fn()
      .mockImplementation((cb) => cb([{ key: "foo", value: "bar" }]));
  });

  it("should return null and not crash if window is undefined", () => {
    jest.useFakeTimers();

    const { result } = renderHook(useFingerprint);

    act(() => {
      jest.runAllTimers();
    });

    expect(result.error).toBeUndefined();
    expect(result.current).toMatchSnapshot();
    expect(result.current).toBe(null);
  });
});
