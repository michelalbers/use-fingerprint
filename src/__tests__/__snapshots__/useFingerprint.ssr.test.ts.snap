// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFingerprint hook should return null and not crash if window is undefined 1`] = `null`;
